Update git-webkit conflict to support additional integration/ branches

akath20 opened the pull request:
#### c2e76f6579ddf2b7c0424c7b23405c1a5c1aff38
<pre>
Update git-webkit conflict to support additional integration/ branches
<a href="https://bugs.webkit.org/show_bug.cgi?id=278562">https://bugs.webkit.org/show_bug.cgi?id=278562</a>
<a href="https://rdar.apple.com/129681381">rdar://129681381</a>

Reviewed by Ryan Haddad.

This will now search for updated integration branches when running git-webkit conflict.

* Tools/Scripts/libraries/webkitscmpy/webkitscmpy/program/conflict.py:
(Conflict):
(Conflict.parser):
(Conflict.find_conflict_pr):
(Conflict.get_open_integration_prs):
(Conflict.main):
* Tools/Scripts/libraries/webkitscmpy/webkitscmpy/test/conflict_unittest.py:
(TestConflict._mock_radar_response):
(TestConflict.test_conflict_not_found):
(TestConflict.test_conflict_found):

Canonical link: <a href="https://commits.webkit.org/283569@main">https://commits.webkit.org/283569@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/0c6076a1e6e265e993266becf51c643e2f5e12ee

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/66676 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/48/builds/46051 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/55/builds/19298 "Built successfully") | [  ~~🛠 wpe~~](https://ews-build.webkit.org/#/builders/5/builds/70710 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🛠 wincairo~~](https://ews-build.webkit.org/#/builders/59/builds/17809 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/68794 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/49/builds/53850 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/61/builds/17575 "Built successfully") | [  ~~🧪 wpe-wk2~~](https://ews-build.webkit.org/#/builders/5/builds/70710 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 wincairo-tests~~](https://ews-build.webkit.org/#/builders/59/builds/17809 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/69743 "Passed tests") | [  ~~🧪 ios-wk2~~](https://ews-build.webkit.org/#/builders/47/builds/42396 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 api-mac~~](https://ews-build.webkit.org/#/builders/18/builds/57700 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 api-wpe~~](https://ews-build.webkit.org/#/builders/5/builds/70710 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| [✅ 🧪 webkitpy](https://ews-build.webkit.org/#/builders/6/builds/66188 "Passed tests") | [  ~~🧪 ios-wk2-wpt~~](https://ews-build.webkit.org/#/builders/42/builds/39067 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 mac-wk1~~](https://ews-build.webkit.org/#/builders/64/builds/15091 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🛠 wpe-cairo~~](https://ews-build.webkit.org/#/builders/65/builds/16163 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| | [  ~~🧪 api-ios~~](https://ews-build.webkit.org/#/builders/13/builds/60981 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 mac-wk2~~](https://ews-build.webkit.org/#/builders/63/builds/15432 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/72412 "Built successfully") | 
| | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/87/builds/10633 "Built successfully") | [  ~~🧪 mac-AS-debug-wk2~~](https://ews-build.webkit.org/#/builders/62/builds/14792 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 gtk-wk2~~](https://ews-build.webkit.org/#/builders/1/builds/60759 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| [✅ 🧪 services](https://ews-build.webkit.org/#/builders/28/builds/66356 "Passed tests") | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/86/builds/10665 "Built successfully") | [  ~~🧪 mac-wk2-stress~~](https://ews-build.webkit.org/#/builders/8/builds/57756 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 api-gtk~~](https://ews-build.webkit.org/#/builders/21/builds/61094 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| | [  ~~🧪 vision-wk2~~](https://ews-build.webkit.org/#/builders/88/builds/8753 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 mac-intel-wk2~~](https://ews-build.webkit.org/#/builders/119/builds/2371 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | 
| [✅ 🛠 🧪 unsafe-merge](https://ews-build.webkit.org/#/builders/22/builds/10114 "Built successfully and passed tests") | [  ~~🛠 tv~~](https://ews-build.webkit.org/#/builders/44/builds/41858 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | | 
| | [  ~~🛠 tv-sim~~](https://ews-build.webkit.org/#/builders/46/builds/42935 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/43/builds/44118 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/45/builds/42678 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

bravc commented on Tools/Scripts/libraries/webkitscmpy/webkitscmpy/program/conflict.py:
```diff
@@ -39,9 +44,28 @@ def parser(cls, parser, loggers=None):
+
+        integration_branches = []
+        for prefix in ('ci', 'conflict'):
+            integration_branches.append(cls.INTEGRATION_BRANCH_PREFIX + "/" + prefix + "/" + shas[0][:10] + "_" + shas[-1][:10])
```
.format string here?

akath20 replied:
I think overall with opensource stuff we have tried to refrain from those python3 specifics things, unless that has changed recently @JonWBedard @briannafan

bravc replied:
.format is not a python3 thing, it is used lower in this file and in many others

briannafan replied:
I think format strings are fine. f-strings are the python3 specific thing you're probably thinking of.

bravc commented on Tools/Scripts/libraries/webkitscmpy/webkitscmpy/program/conflict.py:
```diff
@@ -39,9 +44,28 @@ def parser(cls, parser, loggers=None):
+
+        for entry in radar.sourceChanges.splitlines():
+            repo, action, sha = entry.split(', ')
+            if repo == 'WebKit':
```
To stay generic the repo name may need to be calculated from git information. Other places use `repo_name = rmt.name if '/' not in name else name.split('/', 1)[-1]`

akath20 replied:
messaged cam offline, I think we want hardcoded webkit here since that's what we put in the sourceChanges field.

bravc reviewed (approved):
LGTM, can we get a new test for this case as well?

drobson1005 reviewed (approved):
rs=me

JonWBedard commented on Tools/Scripts/libraries/webkitscmpy/webkitscmpy/program/conflict.py:
```diff
@@ -22,13 +22,15 @@
 class Conflict(Command):
     name = 'conflict'
     help = "Given the representative Radar ID of a conflicting merge, checkout the branch."
+    INTEGRATION_BRANCH_PREFIX = 'integration'
```
cc @briannafan  for thoughts on this, I wonder if this prefix should actually be owned by the repository object in case projects want to vary it. Probably not worth making that change for now.

briannafan replied:
Yeah I'm fine with leaving as is and moving it if it changes.

JonWBedard commented on Tools/Scripts/libraries/webkitscmpy/webkitscmpy/program/conflict.py:
```diff
@@ -39,9 +41,40 @@ def parser(cls, parser, loggers=None):
+        we need to search for PRs with branches that start with the integration prefix
+        and have the first and last sha.
+        """
+        radar_obj = cls.radar_for_id(radar_id)
```
It doesn't seem like we actually need a radar here, right? We just need any issue that we can extract commits from.

briannafan replied:
We can avoid using a new function and calling radar client directly by doing `radar_obj = Tracker.from_string(radar_id)`. This would also work if a bugzilla is passed in. If we want to support bugzilla we can just change the variable names.

JonWBedard commented on Tools/Scripts/libraries/webkitscmpy/webkitscmpy/program/conflict.py:
```diff
@@ -39,9 +41,40 @@ def parser(cls, parser, loggers=None):
+
+        integration_branches = []
+        for prefix in ('ci', 'conflict'):
+            integration_branches.append("{}/{}/{}_{}".format(cls.INTEGRATION_BRANCH_PREFIX, prefix, shas[0][:10], shas[-1][:10]))
```
We should be using 12 character shas, enshrined in the `Commit.HASH_LABEL_SIZE` variable (there's a whole explanation as to why, the TLDR is "birthday paradox, WebKit needs 12 character shas to guarantee uniqueness".

akath20 replied:
ah that's right, let me update that here and in our automerge code

akath20 replied:
https://stashweb.sd.apple.com/projects/SAFARI/repos/automerge/pull-requests/1725/overview

JonWBedard commented on Tools/Scripts/libraries/webkitscmpy/webkitscmpy/program/conflict.py:
```diff
@@ -39,9 +41,40 @@ def parser(cls, parser, loggers=None):
+        shas = []
+
+        repo_name = remote.name if '/' not in remote.name else remote.name.split('/', 1)[-1]
+        for entry in radar_obj.sourceChanges.splitlines():
```
I'll let @briannafan chime in, but this sounds like a case where we should be abstracting "sourceChanges" on issue trackers more generally. I can see a world where, for example, we feed this same command a Bugzilla instead.

At the moment, the best implementation of this probably involves just scraping Bugzilla comments for commit annotations. Also, this is more accurately returning refs, I think. It seems possible, for example, that we would get an identifier instead of a hash. The below logic should handle that just fine since you're basically looking for a PR that matches any 1 of an enumerated pattern.

akath20 replied:
While this is in open source code, the conflict workflow is intended for internal users. Do you think it makes sense to work that into this PR, or should we proceed with as-is for now and reevaluate if needed at a later time?

briannafan replied:
Supporting `sourceChanges` for bugzilla would require significant changes - in the spirit of having this tool available, I think this can come at a later time. That said, we can clean up the code here to use the `Tracker` and `Issue` classes to make that transition easier (see my other comments)

briannafan commented on Tools/Scripts/libraries/webkitscmpy/webkitscmpy/program/conflict.py:
```diff
@@ -39,9 +42,40 @@ def parser(cls, parser, loggers=None):
+        shas = []
+
+        repo_name = remote.name if '/' not in remote.name else remote.name.split('/', 1)[-1]
+        for entry in radar_obj.sourceChanges.splitlines():
```
I think `radar_obj.source_changes` can give you a list right off the bat, assuming you use `Tracker.from_string` to create an issue object

briannafan commented on Tools/Scripts/libraries/webkitscmpy/webkitscmpy/program/conflict.py:
```diff
@@ -22,13 +22,16 @@
+    INTEGRATION_BRANCH_PREFIX = 'integration'
 
     @classmethod
     def parser(cls, parser, loggers=None):
```
It might be a quality of life improvement to change `radar_id` to a string, so you can input anything in the form of `rdar://problem/123456789`, `rdar://123456789`, etc.

akath20 replied:
updated!

briannafan commented on Tools/Scripts/libraries/webkitscmpy/webkitscmpy/program/conflict.py:
```diff
@@ -22,26 +22,64 @@
+                return tracker.client
+
+    @classmethod
+    def radar_for_id(cls, radar_id):
```
We can delete this method since it's not being used anywhere right?

briannafan commented on Tools/Scripts/libraries/webkitscmpy/webkitscmpy/program/conflict.py:
```diff
@@ -52,11 +90,11 @@ def main(cls, args, repository, **kwargs):
             return 1
 
-        expected_branch = 'integration/conflict/{}'.format(args.radar_id)
+        radar_id = int(cls.get_radar_client().extract_radar_ids_from_text(args.radar)[0])
```
If you constructed `radar_obj = Tracker.from_string(f'rdar://{args.radar}')` in main, you can extract the radar id as `radar_obj.id` and avoid the call to radar client! We can remove both `get_radar_client ` and `radar_for_id `.

akath20 replied:
great point, updated!

akath20 committed:
Update git-webkit conflict to support additional integration/ branches
https://bugs.webkit.org/show_bug.cgi?id=278562
rdar://129681381

Reviewed by Ryan Haddad.

This will now search for updated integration branches when running git-webkit conflict.

* Tools/Scripts/libraries/webkitscmpy/webkitscmpy/program/conflict.py:
(Conflict):
(Conflict.parser):
(Conflict.find_conflict_pr):
(Conflict.get_open_integration_prs):
(Conflict.main):
* Tools/Scripts/libraries/webkitscmpy/webkitscmpy/test/conflict_unittest.py:
(TestConflict._mock_radar_response):
(TestConflict.test_conflict_not_found):
(TestConflict.test_conflict_found):

Canonical link: https://commits.webkit.org/283569@main
